core/src/main/java/com/netflix/iceberg/BaseMetastoreTableOperations.java

```diff
@@ -114,30 +103,19 @@ protected void refreshFromMetadataLocation(String newLocation, int numRetries) {
     this.shouldRefresh = false;
   }
 
-  private String metadataFileLocation(TableMetadata metadata, String filename) {
-    String metadataLocation = metadata.properties()
-        .get(TableProperties.WRITE_METADATA_LOCATION);
-
-    if (metadataLocation != null) {
-      return String.format("%s/%s", metadataLocation, filename);
-    } else {
-      return String.format("%s/%s/%s", metadata.location(), METADATA_FOLDER_NAME, filename);
-    }
-  }
-
   @Override
-  public String metadataFileLocation(String filename) {
-    return metadataFileLocation(current(), filename);
+  public FileIO io() {
+    return new HadoopFileIO(conf, currentMetadata);
   }
 
   @Override
-  public FileIO io() {
-    return fileIo;
+  public FileIO io(TableMetadata tableMetadata) {
```

Comment:
Is it possible to make this private? It is odd for a TableOperations class that provides TableMetadata instances to have a public method that accepts TableMetadata instead of using current.

Reply:
Sorry, I'm now looking at the javadoc for it.

Reply:
Okay, looks like the trade-off here is that BaseTransaction needs to provide a FileIO object and its view of TableMetadata may be different from the table's actual metadata state because we stack up multiple changes and then commit just the final version in the transaction.

So unless I'm missing something again, the only thing that adding this method accomplishes is to be able to update some location and have the change take effect for a write in the same atomic commit. I'm okay with that limitation to make this API cleaner, unless you need to be able to change settings and write in an atomic commit for your use case.

Reply:
> So unless I'm missing something again, the only thing that adding this method accomplishes is to be able to update some location and have the change take effect for a write in the same atomic commit. I'm okay with that limitation to make this API cleaner, unless you need to be able to change settings and write in an atomic commit for your use case.

So I accidentally stumbled upon having to do it this way because a lot of tests just failed if we used the old location to commit when the new metadata specifies committing elsewhere. Thus I had thought that the current behavior is part of public API and we don't intend on changing it. I suppose we ought to document this specific behavior somewhere (if you change the new metadata location table property on a commit, the change isn't reflected in that current commit). However it's unclear where the right place to document that is.